lib/math/math.go

```diff
@@ -0,0 +1,150 @@
+//                   The point of atan2() is that the signs of both inputs are known to it, so it can compute the correct
+//                   quadrant for the angle.
+//                   For example, atan(1) and atan2(1, 1) are both pi/4, but atan2(-1, -1) is -3*pi/4.
+//     cos(x) - Returns the cosine of x, in radians.
+//     hypot(x, y) - Returns the Euclidean norm, sqrt(x*x + y*y). This is the length of the vector from the origin to point (x, y).
+//     sin(x) - Returns the sine of x, in radians.
+//     tan(x) - Returns the tangent of x, in radians.
+//
+//     degrees(x) - Converts angle x from radians to degrees.
+//     radians(x) - Converts angle x from degrees to radians.
+//
+//     acosh(x) - Returns the inverse hyperbolic cosine of x.
+//     asinh(x) - Returns the inverse hyperbolic sine of x.
+//     atanh(x) - Returns the inverse hyperbolic tangent of x.
+//     cosh(x) - Returns the hyperbolic cosine of x.
+//     sinh(x) - Returns the hyperbolic sine of x.
+//     tanh(x) - Returns the hyperbolic tangent of x.
+//
+//     log(x) - Returns the natural logarithm of x.
+//     log2(x) - Returns the decimal logarithm of x. The special cases are the same as for Log.
+//     log10(x) - Returns the natural logarithm of 1 plus its argument x. It is more accurate than Log(1 + x) when x is near.
+//
+// All function accept both int and float values as arguments.
+//
+// The module also defines approximations of the following constants:
+//
+//     e - The base of natural logarithms, approximately 2.71828.
+//     phi - The golden ratio, (1 + sqrt(5)) / 2, approximately 1.61803.
+//     pi - The ratio of a circle's circumference to its diameter, approximately 3.14159.
+//
+var Module = &starlarkstruct.Module{
+	Name: "math",
+	Members: starlark.StringDict{
+		"abs":   newUnaryBuiltin("abs", math.Abs),
+		"ceil":  newUnaryBuiltin("ceil", math.Ceil),
+		"floor": newUnaryBuiltin("floor", math.Floor),
+		"round": newUnaryBuiltin("round", math.Round),
+
+		"exp":  newUnaryBuiltin("exp", math.Exp),
+		"sqrt": newUnaryBuiltin("sqrt", math.Sqrt),
+
+		"acos":  newUnaryBuiltin("acos", math.Acos),
+		"asin":  newUnaryBuiltin("asin", math.Asin),
+		"atan":  newUnaryBuiltin("atan", math.Atan),
+		"atan2": newBinaryFunction("atan2", math.Atan2),
+		"cos":   newUnaryBuiltin("cos", math.Cos),
+		"hypot": newBinaryFunction("hypot", math.Hypot),
+		"sin":   newUnaryBuiltin("sin", math.Sin),
+		"tan":   newUnaryBuiltin("tan", math.Tan),
+
+		"degrees": newUnaryBuiltin("degrees", degrees),
+		"radians": newUnaryBuiltin("radians", radians),
+
+		"acosh": newUnaryBuiltin("acosh", math.Acosh),
+		"asinh": newUnaryBuiltin("asinh", math.Asinh),
+		"atanh": newUnaryBuiltin("atanh", math.Atanh),
+		"cosh":  newUnaryBuiltin("cosh", math.Cosh),
+		"sinh":  newUnaryBuiltin("sinh", math.Sinh),
+		"tanh":  newUnaryBuiltin("tanh", math.Tanh),
+
+		"log":   newUnaryBuiltin("log", math.Log),
```

Comment:
Python's math.log function accepts an optional base argument (default=e), and so should Starlark's. The body of the function is `math.Log(x) / math.Log(base)`, but you'll need to specialize newBinaryFunction because of the optional argument.

Reply:
Please expand out the implementation of log in full, and don't complicate all the other functions.